c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -0,0 +1,392 @@
+            }
+        }
+    }
+
+    /// Get subexpression `idx` of `ex`.  Panics if the index is out of range for `ex`.  The
+    /// precise meaning of the index depends on the expression kind.
+    fn get_subexpr(&self, ex: &'tcx hir::Expr<'tcx>, idx: usize) -> Rewrite {
+        use hir::ExprKind::*;
+        let sub_ex = match (&ex.kind, idx) {
+            (&Box(e), 0) => e,
+            (&Array(es), i) => &es[i],
+            (&Call(_, args), i) => &args[i],
+            (&MethodCall(_, args, _), i) => &args[i],
+            (&Tup(es), i) => &es[i],
+            (&Binary(_, x, _), 0) => x,
+            (&Binary(_, _, y), 1) => y,
+            (&Unary(_, x), 0) => x,
+            (&Cast(e, _), 0) => e,
+            (&Type(e, _), 0) => e,
+            (&DropTemps(e), 0) => e,
+            (&If(cond, _, _), 0) => cond,
+            (&If(_, then, _), 1) => then,
+            (&If(_, _, Some(else_)), 2) => else_,
+            (&Match(e, _, _), 0) => e,
+            (&Assign(l, _, _), 0) => l,
+            (&Assign(_, r, _), 1) => r,
+            (&AssignOp(_, l, _), 0) => l,
+            (&AssignOp(_, _, r), 1) => r,
+            (&Field(e, _), 0) => e,
+            (&Index(arr, _), 0) => arr,
+            (&Index(_, e_idx), 1) => e_idx,
+            (&AddrOf(_, _, e), 0) => e,
+            (&Break(_, Some(e)), 0) => e,
+            (&Ret(Some(e)), 0) => e,
+            (&Struct(_, flds, base), i) => {
+                if i == flds.len() {
+                    base.unwrap()
+                } else {
+                    flds[i].expr
+                }
+            }
+            (&Repeat(e, _), 0) => e,
+            (&Yield(e, _), 0) => e,
+            _ => panic!("bad subexpression index {} for {:?}", idx, ex),
+        };
+        Rewrite::Sub(idx, sub_ex.span)
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.tcx.hir()
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        let mut hir_rw = Rewrite::Identity;
+
+        if let Some(loc) = self.find_primary_location(ex) {
+            self.locations_visited.insert(loc);
+            let rws = self.rewrites.get(&loc).map_or(&[] as &[_], |v| v);
+            for rw in rws {
+                match rw.kind {
+                    mir_op::RewriteKind::OffsetSlice { mutbl } => {
+                        assert!(matches!(hir_rw, Rewrite::Identity));
+                        //assert_eq!(num_args, 2);
+                        // `p.offset(i)` -> `&p[i as usize ..]`
+                        let arr = self.get_subexpr(ex, 0);
+                        let idx = Rewrite::CastUsize(Box::new(self.get_subexpr(ex, 1)));
+                        let elem = Rewrite::SliceTail(Box::new(arr), Box::new(idx));
+                        hir_rw = Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl));
+                    }
```

Comment:
For this `match`, can you hoist the `hir_rw = ` out of the `match` so it's clear it's unconditionally re-assigned?

Reply:
Actually, can you extract the whole `match` into its own method?

Reply:
What about `mir_to_hir_rewrite`?